pipeline/utils/github_utils.py

```diff
@@ -0,0 +1,137 @@
+# Copyright 2016 Google Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Utility functions related to GitHub interaction. Uses the GitHub API with
+reference https://developer.github.com/v3/"""
+
+# How to update requirements so that these imports are possible?
+import requests
+import json
+import os
+import base64
+
+
+_API_URL = 'https://api.github.com'
+
+# file modes for different git file types
+_TREE_EL_MODES = {'file': u'100644',
+                  'executable': u'100755',
+                  'tree': u'040000',
+                  'symlink': u'120000'}
+
+
+def _is_file(el):
+    return os.path.isfile(el) and not os.access(el, os.X_OK)
+
+
+def _is_executable(el):
+    return os.path.isfile(el) and os.access(el, os.X_OK)
+
+
+def _get_entries(entry_type_func, a_dir):
+    """List entries in a_dir that are of type specified by entry_type_func"""
+    return filter(lambda x: entry_type_func(os.path.join(a_dir, x)),
+                  os.listdir(a_dir))
+
+
+def _get_request_content(req_func, url, req_data=None, username=None,
+                         password=None):
+    """Make an HTTP request with given data and (username, password) tuple.
+    Throw an exception if the request fails"""
+    if username and password:
+        attempt = req_func(url, data=req_data, auth=(username, password))
+    else:
+        attempt = req_func(url, data=req_data)
+    if attempt.ok:
+        return json.loads(attempt.content)
+    else:
+        raise Exception('Failed request at {0}.'
+                        'Reason: {1}'.format(url, attempt.content))
+
+
+def _build_git_tree_blob_item(file_or_exec, blob_mode, curr_dir, username,
```

Comment:
GitHub requires blobs to have different modes if they are executables rather than files.